service/src/main/java/com/epam/esm/services/service/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,126 @@
+package com.epam.esm.services.service;
+
+import com.epam.esm.entities.GiftCertificate;
+import com.epam.esm.entities.Tag;
+import com.epam.esm.repository.GiftCertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.services.dto.GiftCertificateDto;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+@Transactional
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+    private final GiftCertificateRepository giftCertificateRepository;
+    private final TagRepository tagRepository;
+    private static final String GIFT_CERTIFICATES_SORTING_CONDITION_BY_DATE = "date";
+    private static final String GIFT_CERTIFICATES_SORTING_CONDITION_BY_DATE_DESC = "date-desc";
+    private static final String GIFT_CERTIFICATES_SORTING_CONDITION_BY_NAME = "name";
+    private static final String GIFT_CERTIFICATES_SORTING_CONDITION_BY_NAME_DESC = "name-desc";
```

Comment:
better to have separate parameters for sorting direction and field name. Now if you want to add **one** field to sort you have to add **two** new parameters.